components/builder-depot/src/server.rs

```diff
@@ -841,9 +857,21 @@ fn upload_package(req: &mut Request) -> IronResult<Response> {
             }
         };
 
-    if origin_package_found && depot.archive(&ident, &target_from_artifact).is_some() {
-        return Ok(Response::with(status::Conflict));
-    };
+    // If an S3 backend is set in the config file
+    // check whether the package already exists in
+    // the configured bucket, otherwise validate on disk
+    if origin_package_found {
+        let s3_config = depot.config.s3.to_owned();
```

Comment:
I see this is used in a few places. Would it make sense to move the S3 connection into the global context that is passed around with the request handers?